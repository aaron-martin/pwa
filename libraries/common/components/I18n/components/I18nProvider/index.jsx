/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import {
  getTranslator,
  getPriceFormatter,
  getDateFormatter,
  getTimeFormatter,
} from '../../../../helpers/i18n';

/**
 * A React component that provides child components with i18n features.
 * @returns {React.Component}
 */
export default class I18nProvider extends React.Component {
  static propTypes = {
    lang: PropTypes.string.isRequired,
    locales: PropTypes.shape().isRequired,
    children: PropTypes.node,
  };

  static defaultProps = {
    children: null,
  };

  static childContextTypes = {
    i18n: PropTypes.func.isRequired,
  };

  /**
   * Provides context for child components.
   * @returns {Object}
   */
  getChildContext() {
    return {
      i18n: this.getI18nInstance,
    };
  }

  /**
   * Gets a shortcut to i18n functionality with preset language.
   * @returns {Object}
   */
  getI18nInstance = () => {
    const { locales, lang } = this.props;

    return {
      __: getTranslator(locales, lang),
      _p: getPriceFormatter(lang),
      _d: getDateFormatter(lang),
      _t: getTimeFormatter(lang),
    };
  };

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const { children } = this.props;

    return children
      ? React.cloneElement(children)
      : null;
  }
}
