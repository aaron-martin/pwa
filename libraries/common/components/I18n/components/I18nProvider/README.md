# I18nProvider
---
The I18nProvider component provides access to i18n functionalities via `context.i18n()`
to all child components.

## Getting Started
```markup
import { I18nProvider } from 'Library/components';

<I18nProvider
  lang="en_US"
  locales={languageLocales}
>
  <MyAppComponent />
</I18nProvider>
```

## Props

### lang (required)

_Type_: `string`

A string representing the current language code.

E.g. "de_DE", "de_AT", "en_US", ...

###### Usage:

```
<I18nProvider lang="de_DE" ... />
```

### locales (required)

_Type_: `Object`

An object containing the locale information for a specific language.

###### Usage:

```
const locales = {
  en_US: {
    greeting: 'Hello {name}',
  },
};

<I18nProvider locales={locales} ... />
```

### children (optional)

_Type_: `ReactElement`

Only a single child is supported to be passed as `children`.

## Context

### i18n

_Type_: `Function`

A function that returns an object with access to translation functionality
for the currently set language and locales.

The returned object contains the following functions:

- `__(translationKey: string, params: Object)`  
can be used for text translations.

- `_p(price: number, currency: string)`  
can be used for price formatting.

- `_d(timestamp: number, format: string)`  
can be used for date formatting.

- `_t(timestamp: number, format: string)`  
can be used for time formatting.


###### Usage:

```
const MyComponent = (props, context) => {
  const { __ } = context.i18n();
  
  return <div>{__('translationKey')}</div>;
};

<I18nProvider ... />
  <MyComponent />
</I18nProvider>
```
