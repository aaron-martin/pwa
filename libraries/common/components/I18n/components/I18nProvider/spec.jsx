/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow } from 'enzyme';
import I18nProvider from './index';

describe('<I18nProvider />', () => {
  let renderedElement;
  let renderedInstance;
  const testLocales = {
    en: {
      greeting: 'Hello {name}',
    },
    de: {
      greeting: 'Guten Tag {name}',
    },
  };

  /**
   * Renders the component.
   * @param {Object} props The component props.
   */
  const renderComponent = (props) => {
    renderedElement = shallow(<I18nProvider {...props} />);
    renderedInstance = renderedElement.instance();
  };

  beforeEach(() => {
    renderComponent({
      lang: 'en_US',
      locales: testLocales.en,
    });
  });

  describe('Given the component was mounted to the DOM', () => {
    it('should match snapshot', () => {
      expect(renderedElement).toMatchSnapshot();
    });

    it('should provide access to i18n via context', () => {
      const { i18n } = renderedInstance.getChildContext();
      expect(i18n).toEqual(renderedInstance.getI18nInstance);
    });

    it('should translate when calling an instance method', () => {
      const { __ } = renderedInstance.getI18nInstance();
      const translated = __('greeting', { name: 'Test' });
      expect(translated).toBe('Hello Test');
    });

    describe('Given the lang prop changes', () => {
      beforeEach(() => {
        renderedElement.setProps({
          lang: 'de_DE',
          locales: testLocales.de,
        });
      });

      it('should translate according to changed language', () => {
        const { __ } = renderedInstance.getI18nInstance();
        const translated = __('greeting', { name: 'Test' });
        expect(translated).toBe('Guten Tag Test');
      });
    });
  });
});
