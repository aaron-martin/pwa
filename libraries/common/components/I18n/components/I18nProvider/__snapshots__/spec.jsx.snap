// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<I18nProvider /> Given the component was mounted to the DOM should match snapshot 1`] = `null`;
